File: packages/deon-python/source/deon/cache.py
"""The response cache.

Two requirements of specification 9, and they are the reason this is not a dictionary keyed by URL:

> Tokens MUST NOT appear in diagnostics or cache identifiers in plain text.
> Authenticated cache entries MUST be separated by a digest of the credential.

So an entry is keyed by `sha256(name + NUL + token)`. The digest keeps the credential out of the
filename, and folding the token *into* the key is what stops a document fetched under one credential
from being served to the holder of another — which is not a cache miss, it is a data leak.

A cache entry is itself a canonical Deon document. That is a small piece of dogfooding with a real
edge: the format has to survive a round trip, so it is made to, on every write and every read.
"""

from __future__ import annotations

import hashlib
import os
import pathlib
import time
from typing import Optional

from .options import ParseOptions
from .value import DeonMap, Value


DEFAULT_CACHE_DURATION = 3_600_000  # one hour, in milliseconds

DEFAULT_CACHE_DIRECTORY = "~/.deon-cache"


def cache_key(name: str, token: str) -> str:
    """The identity of a cached response.

    The NUL is a separator that cannot occur in either half, so no pair of (name, token) can be spelled
    two ways and collide. Without it, a name ending in a token's prefix would hash to the same place.
    """
    digest = hashlib.sha256()
    digest.update(name.encode("utf-8"))
    digest.update(b"\x00")
    digest.update(token.encode("utf-8"))

    return digest.hexdigest()


def now_milliseconds() -> int:
    return int(time.time() * 1000)


def entry_path(name: str, options: ParseOptions) -> Optional[pathlib.Path]:
    if not options.cache:
        return None

    directory = pathlib.Path(os.path.expanduser(options.cache_directory or DEFAULT_CACHE_DIRECTORY))

    return directory / cache_key(name, options.token)


def read(name: str, options: ParseOptions) -> Optional[Value]:
    """A cached response, if there is one and it has not expired.

    Every failure is silent. A cache that raises is worse than no cache: it turns a performance
    decision into a correctness one, and a document that parsed yesterday would stop parsing because
    of a file nobody meant to be load-bearing.
    """
    from . import parse

    path = entry_path(name, options)

    if path is None:
        return None

    try:
        source = path.read_text("utf-8")
    except OSError:
        return None

    try:
        entry = parse(source)
    except Exception:
        return None

    if not isinstance(entry, DeonMap):
        return None

    try:
        cached_at = int(entry["cachedAt"])
        duration = int(entry["cacheDuration"])
    except (KeyError, TypeError, ValueError):
        return None

    if cached_at + duration < now_milliseconds():
        # Expired, so it is gone. Leaving it would mean reading and re-deciding it every time.
        try:
            path.unlink()
        except OSError:
            pass

        return None

    return entry.get("data")


def write(name: str, value: Value, options: ParseOptions) -> None:
    from .stringifier import canonical

    path = entry_path(name, options)

    if path is None:
        return

    entry = DeonMap()
    entry.insert("cachedAt", str(now_milliseconds()))
    entry.insert("cacheDuration", str(options.cache_duration))
    entry.insert("data", value)

    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(canonical(entry), "utf-8")
    except OSError:
        # Silent, for the same reason as above.
        pass

File: packages/deon-go/cache.go
package deon

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
	"path/filepath"
	"strconv"
	"time"
)

// The response cache. Two requirements of specification 9, and they are the reason this is not a map
// keyed by URL:
//
//	Tokens MUST NOT appear in diagnostics or cache identifiers in plain text.
//	Authenticated cache entries MUST be separated by a digest of the credential.
//
// So an entry is keyed by sha256(name + NUL + token). The digest keeps the credential out of the
// filename, and folding the token into the key is what stops a document fetched under one credential
// from being served to the holder of another — which is not a cache miss, it is a data leak.
//
// A cache entry is itself a canonical Deon document. That is a small piece of dogfooding with a real
// edge: the format has to survive a round trip, so it is made to, on every write and every read.

const (
	defaultCacheDuration  = 3_600_000 // one hour, in milliseconds
	defaultCacheDirectory = "~/.deon-cache"
)

// cacheKey is the identity of a cached response. The NUL is a separator that cannot occur in either
// half, so no pair of (name, token) can be spelled two ways and collide: without it, a name ending in
// a token's prefix would hash to the same place.
func cacheKey(name, token string) string {
	digest := sha256.New()
	digest.Write([]byte(name))
	digest.Write([]byte{0})
	digest.Write([]byte(token))
	return hex.EncodeToString(digest.Sum(nil))
}

func nowMilliseconds() int64 {
	return time.Now().UnixMilli()
}

func cacheDurationOf(options *ParseOptions) int {
	if options.CacheDuration > 0 {
		return options.CacheDuration
	}
	return defaultCacheDuration
}

// cacheEntryPath is where a response is written, or false when caching is off.
func cacheEntryPath(name, token string, options *ParseOptions) (string, bool) {
	if !options.Cache {
		return "", false
	}
	directory := options.CacheDirectory
	if directory == "" {
		directory = defaultCacheDirectory
	}
	directory = expandUser(directory)
	return filepath.Join(directory, cacheKey(name, token)), true
}

// expandUser resolves a leading ~ to the caller's home directory, leaving everything else alone.
func expandUser(path string) string {
	if path == "~" || len(path) >= 2 && path[:2] == "~/" {
		if home, err := os.UserHomeDir(); err == nil {
			if path == "~" {
				return home
			}
			return filepath.Join(home, path[2:])
		}
	}
	return path
}

// cacheRead returns a cached response, if there is one and it has not expired. Every failure is
// silent: a cache that raised would turn a performance decision into a correctness one, and a document
// that parsed yesterday would stop parsing because of a file nobody meant to be load-bearing.
func cacheRead(name, token string, options *ParseOptions) (Value, bool) {
	path, ok := cacheEntryPath(name, token, options)
	if !ok {
		return nil, false
	}
	source, err := os.ReadFile(path)
	if err != nil {
		return nil, false
	}
	entry, err := Parse(string(source))
	if err != nil {
		return nil, false
	}
	entryMap, ok := entry.(*Map)
	if !ok {
		return nil, false
	}

	cachedAt, ok1 := cacheInt(entryMap, "cachedAt")
	duration, ok2 := cacheInt(entryMap, "cacheDuration")
	if !ok1 || !ok2 {
		return nil, false
	}
	if cachedAt+duration < nowMilliseconds() {
		// Expired, so it is gone. Leaving it would mean reading and re-deciding it every time.
		_ = os.Remove(path)
		return nil, false
	}

	data, _ := entryMap.Get("data")
	return data, true
}

// cacheWrite stores a response as a canonical Deon entry. Silent on failure, for the same reason.
func cacheWrite(name, token string, value Value, options *ParseOptions) {
	path, ok := cacheEntryPath(name, token, options)
	if !ok {
		return
	}
	entry := NewMap()
	entry.Set("cachedAt", strconv.FormatInt(nowMilliseconds(), 10))
	entry.Set("cacheDuration", strconv.Itoa(cacheDurationOf(options)))
	entry.Set("data", value)

	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return
	}
	_ = os.WriteFile(path, []byte(canonical(entry)), 0o644)
}

func cacheInt(m *Map, key string) (int64, bool) {
	value, ok := m.Get(key)
	if !ok {
		return 0, false
	}
	text, ok := value.(string)
	if !ok {
		return 0, false
	}
	n, err := strconv.ParseInt(text, 10, 64)
	if err != nil {
		return 0, false
	}
	return n, true
}
